File: spring_java/src/test/java/dev/pollito/spring_java/config/advice/ControllerAdviceTest.java
package dev.pollito.spring_java.config.advice;

import static dev.pollito.spring_java.test.util.ApiResponseMatchers.hasErrorFields;
import static dev.pollito.spring_java.test.util.ApiResponseMatchers.hasStandardApiResponseFields;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;
import static org.springframework.http.HttpMethod.GET;
import static org.springframework.http.HttpStatus.*;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
import static org.springframework.test.web.servlet.setup.MockMvcBuilders.standaloneSetup;

import jakarta.servlet.http.HttpServletRequest;
import jakarta.validation.ConstraintViolationException;
import java.util.Set;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.http.HttpStatus;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.servlet.resource.NoResourceFoundException;

class ControllerAdviceTest {

  private MockMvc mockMvc;
  private final HttpServletRequest request = mock(HttpServletRequest.class);

  @RestController
  @RequestMapping("/fake")
  static class FakeController {

    @GetMapping("/not-found")
    public void throwNoResourceFoundException() throws NoResourceFoundException {
      throw new NoResourceFoundException(GET, "/fake", "no-resource-found");
    }

    @GetMapping("/error")
    public void throwException() throws Exception {
      throw new Exception("Test exception");
    }

    @GetMapping("/bad-request")
    public void throwConstraintViolationException() {
      throw new ConstraintViolationException("Constraint violation", Set.of());
    }
  }

  @BeforeEach
  void setUp() {
    mockMvc =
        standaloneSetup(new FakeController())
            .setControllerAdvice(new ControllerAdvice(request))
            .build();
  }

  @Test
  void whenNoResourceFoundException_thenReturnsNotFound() throws Exception {
    HttpStatus status = NOT_FOUND;
    String expectedInstance = "/fake/not-found";

    when(request.getRequestURI()).thenReturn(expectedInstance);
    mockMvc
        .perform(get(expectedInstance))
        .andExpect(status().isNotFound())
        .andExpect(hasStandardApiResponseFields(expectedInstance, status))
        .andExpect(hasErrorFields(status));
  }

  @Test
  void whenException_thenReturnsInternalServerError() throws Exception {
    HttpStatus status = INTERNAL_SERVER_ERROR;
    String expectedInstance = "/fake/error";

    when(request.getRequestURI()).thenReturn(expectedInstance);
    mockMvc
        .perform(get(expectedInstance))
        .andExpect(status().isInternalServerError())
        .andExpect(hasStandardApiResponseFields(expectedInstance, status))
        .andExpect(hasErrorFields(status));
  }

  @Test
  void whenConstraintViolationException_thenReturnsBadRequest() throws Exception {
    HttpStatus status = BAD_REQUEST;
    String expectedInstance = "/fake/bad-request";

    when(request.getRequestURI()).thenReturn(expectedInstance);
    mockMvc
        .perform(get(expectedInstance))
        .andExpect(status().isBadRequest())
        .andExpect(hasStandardApiResponseFields(expectedInstance, status))
        .andExpect(hasErrorFields(status));
  }
}

File: spring_kotlin/src/test/kotlin/dev/pollito/spring_kotlin/config/advice/ControllerAdviceTest.kt
package dev.pollito.spring_kotlin.config.advice

import dev.pollito.spring_kotlin.test.util.hasErrorFields
import dev.pollito.spring_kotlin.test.util.hasStandardApiResponseFields
import io.mockk.every
import io.mockk.mockk
import jakarta.servlet.http.HttpServletRequest
import jakarta.validation.ConstraintViolationException
import kotlin.test.Test
import org.junit.jupiter.api.BeforeEach
import org.springframework.http.HttpMethod.GET
import org.springframework.http.HttpStatus.*
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.get
import org.springframework.test.web.servlet.setup.MockMvcBuilders.standaloneSetup
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.servlet.resource.NoResourceFoundException

class ControllerAdviceTest {

  private lateinit var mockMvc: MockMvc
  private val request: HttpServletRequest = mockk(relaxed = true)

  @RestController
  @RequestMapping("/fake")
  class FakeController {

    @GetMapping("/not-found")
    fun throwNoResourceFoundException() {
      throw NoResourceFoundException(GET, "/fake/not-found", "no-resource-found")
    }

    @GetMapping("/error")
    fun throwException() {
      throw Exception("Test exception")
    }

    @GetMapping("/bad-request")
    fun throwConstraintViolationException() {
      throw ConstraintViolationException("Constraint violation", emptySet())
    }
  }

  @BeforeEach
  fun setUp() {
    mockMvc =
        standaloneSetup(FakeController()).setControllerAdvice(ControllerAdvice(request)).build()
  }

  @Test
  fun `when NoResourceFoundException then returns NotFound`() {
    val status = NOT_FOUND
    val expectedInstance = "/fake/not-found"

    every { request.requestURI } returns expectedInstance

    mockMvc.get(expectedInstance).andExpect {
      status { isNotFound() }
      hasStandardApiResponseFields(expectedInstance, status)
      hasErrorFields(status)
    }
  }

  @Test
  fun `when Exception then returns InternalServerError`() {
    val status = INTERNAL_SERVER_ERROR
    val expectedInstance = "/fake/error"

    every { request.requestURI } returns expectedInstance

    mockMvc.get(expectedInstance).andExpect {
      status { isInternalServerError() }
      hasStandardApiResponseFields(expectedInstance, status)
      hasErrorFields(status)
    }
  }

  @Test
  fun `when ConstraintViolationException then returns BadRequest`() {
    val status = BAD_REQUEST
    val expectedInstance = "/fake/bad-request"

    every { request.requestURI } returns expectedInstance

    mockMvc.get(expectedInstance).andExpect {
      status { isBadRequest() }
      hasStandardApiResponseFields(expectedInstance, status)
      hasErrorFields(status)
    }
  }
}
